river/Layout.zig

```diff
@@ -0,0 +1,182 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2020 - 2021 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const std = @import("std");
+const wlr = @import("wlroots");
+const wayland = @import("wayland");
+const wl = wayland.server.wl;
+const zriver = wayland.server.zriver;
+
+const log = @import("log.zig");
+const util = @import("util.zig");
+
+const Box = @import("Box.zig");
+const Server = @import("Server.zig");
+const Output = @import("Output.zig");
+const View = @import("View.zig");
+const ViewStack = @import("view_stack.zig").ViewStack;
+const LayoutDemand = @import("LayoutDemand.zig");
+
+layout: *zriver.LayoutV1,
+namespace: ?[]const u8,
+output: *Output,
+
+pub fn init(self: *Self, namespace: [*:0]const u8, output: *Output, client: *wl.Client, version: u32, id: u32) !void {
+    self.output = output;
+    self.namespace = try util.gpa.dupe(u8, std.mem.span(namespace));
```

Comment:
this gets leaked if the next line returns an error. You could use errdefer here or postpone allocation till later and use `wl.Client.postNoMemory()` on failure